src/java/main/org/apache/zookeeper/server/WatchManager.java

```diff
@@ -181,11 +206,18 @@ synchronized void dumpWatches(PrintWriter pwriter, boolean byPath) {
      * @return true if the watcher exists, false otherwise
      */
     synchronized boolean containsWatcher(String path, Watcher watcher) {
-        HashSet<String> paths = watch2Paths.get(watcher);
-        if (paths == null || !paths.contains(path)) {
-            return false;
+        PathIterator pathIterator = new PathIterator(path);
+        for (String localPath : pathIterator.asIterable()) {
+            Map<Watcher, Type> watchers = watchTable.get(localPath);
+            Type watcherType = (watchers != null) ? watchers.get(watcher) : null;
+            if ((watcherType == Type.STANDARD) && !pathIterator.atParentPath()) {
```

Comment:
NIT: I feel this is likely to be more performant if we short circuit on atParentPath.